src/audio/mixin_mixout/mixin_mixout.c

```diff
@@ -727,13 +718,95 @@ static int mixout_prepare(struct processing_module *mod,
 	 * Since mixout sink buffer stream is reset on .prepare(), let's
 	 * reset counters for not yet produced frames in that buffer.
 	 */
-	mod_source_info = module_source_info_acquire(mod->source_info);
-	md = mod_source_info->private;
+	md = module_get_private_data(mod);
 	md->mixed_frames = 0;
 
 	for (i = 0; i < MIXOUT_MAX_SOURCES; i++)
-		md->pending_frames[i] = 0;
-	module_source_info_release(mod_source_info);
+		md->pending_frames[i].frames = 0;
+
+	return 0;
+}
+
+int mixout_bind(struct processing_module *mod, void *data)
+{
+	struct ipc4_module_bind_unbind *bu;
+	struct comp_dev *mixin;
+	struct pending_frames *pending_frames;
+	int src_id;
+	struct mixout_data *mixout_data;
+
+	comp_dbg(mod->dev, "mixout_bind() %p", data);
+
+	bu = (struct ipc4_module_bind_unbind *)data;
+	src_id = IPC4_COMP_ID(bu->primary.r.module_id, bu->primary.r.instance_id);
+
+	/* we are only interested in bind for mixin -> mixout pair */
+	if (mod->dev->ipc_config.id == src_id)
+		return 0;
+
+	mixin = ipc4_get_comp_dev(src_id);
+	if (!mixin) {
+		comp_err(mod->dev, "mixout_bind: no mixin with ID %d found", src_id);
+		return -EINVAL;
+	}
+
+	mixout_data = module_get_private_data(mod);
+
+	pending_frames = get_mixin_pending_frames(mixout_data, mixin);
+	/*
+	 * this should never happen as pending_frames info for a particular mixin and mixout pair
+	 * should have been already cleared in mixout_unbind()
+	 */
+	if (pending_frames) {
+		pending_frames->mixin = NULL;
+		pending_frames->frames = 0;
+	}
+
+	/* find an empty slot in the pending_frames array */
+	pending_frames = get_mixin_pending_frames(mixout_data, NULL);
+	if (!pending_frames) {
+		/* no free slot in pending_frames array */
+		comp_err(mod->dev, "Too many inputs!");
+		return -ENOMEM;
+	}
+
+	pending_frames->mixin = mixin;
+	pending_frames->frames = 0;
```

Comment:
you need to swap these: if between them the copying thread finds this mixin with uninitialised `.frames` - it will try to copy data from it...

Reply:
Swapped. However, this was not really necessary:
(1) frames are cleared in unbind() and on initialization,
(2) bind() is executed on same core as LL and it blocks interrupts (blocks LL to preempt IPC or IDC thread). So bind cannot be preempted by LL (at least this is true for same core binding). This version of mixin and mixout only supports same core connection. For cross-core connection (we are still uncertain if we need it) a separate version of mixin and mixout should be implemented.